// expo-cli/build/commands/eas-build/build/builders/AndroidBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _easBuildJob() {
  const data = require("@expo/eas-build-job");

  _easBuildJob = function () {
    return data;
  };

  return data;
}

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function () {
    return data;
  };

  return data;
}

function _figures() {
  const data = _interopRequireDefault(require("figures"));

  _figures = function () {
    return data;
  };

  return data;
}

function _fsExtra() {
  const data = _interopRequireDefault(require("fs-extra"));

  _fsExtra = function () {
    return data;
  };

  return data;
}

function _ora() {
  const data = _interopRequireDefault(require("ora"));

  _ora = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _CommandError() {
  const data = _interopRequireDefault(require("../../../../CommandError"));

  _CommandError = function () {
    return data;
  };

  return data;
}

function _read() {
  const data = require("../../../../credentials/credentialsJson/read");

  _read = function () {
    return data;
  };

  return data;
}

function _AndroidCredentialsProvider() {
  const data = _interopRequireDefault(require("../../../../credentials/provider/AndroidCredentialsProvider"));

  _AndroidCredentialsProvider = function () {
    return data;
  };

  return data;
}

function _easJson() {
  const data = require("../../../../easJson");

  _easJson = function () {
    return data;
  };

  return data;
}

function _git() {
  const data = require("../../../../git");

  _git = function () {
    return data;
  };

  return data;
}

function _log() {
  const data = _interopRequireDefault(require("../../../../log"));

  _log = function () {
    return data;
  };

  return data;
}

function gitUtils() {
  const data = _interopRequireWildcard(require("../../utils/git"));

  gitUtils = function () {
    return data;
  };

  return data;
}

function _credentials() {
  const data = require("../credentials");

  _credentials = function () {
    return data;
  };

  return data;
}

function _gradleContent() {
  const data = _interopRequireDefault(require("../templates/gradleContent"));

  _gradleContent = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class AndroidBuilder {
  constructor(ctx) {
    this.ctx = ctx;

    _defineProperty(this, "credentials", void 0);

    _defineProperty(this, "secretEnvs", void 0);

    _defineProperty(this, "credentialsPrepared", false);
  }

  async setupAsync() {}

  async ensureCredentialsAsync() {
    this.credentialsPrepared = true;
    this.secretEnvs = await (0, _read().readSecretEnvsAsync)(this.ctx.commandCtx.projectDir);

    if (!this.shouldLoadCredentials()) {
      return;
    }

    const provider = new (_AndroidCredentialsProvider().default)(this.ctx.commandCtx.projectDir, {
      projectName: this.ctx.commandCtx.projectName,
      accountName: this.ctx.commandCtx.accountName
    }, {
      nonInteractive: this.ctx.commandCtx.nonInteractive
    });
    await provider.initAsync();
    const credentialsSource = await (0, _credentials().ensureCredentialsAsync)(provider, this.ctx.buildProfile.workflow, this.ctx.buildProfile.credentialsSource, this.ctx.commandCtx.nonInteractive);
    this.credentials = await provider.getCredentialsAsync(credentialsSource);
    return credentialsSource;
  }

  async isProjectConfiguredAsync() {
    const androidAppDir = _path().default.join(this.ctx.commandCtx.projectDir, 'android', 'app');

    const buildGradlePath = _path().default.join(androidAppDir, 'build.gradle');

    const easGradlePath = _path().default.join(androidAppDir, 'eas-build.gradle');

    const hasEasGradleFile = await _fsExtra().default.pathExists(easGradlePath);
    const buildGradleContent = await _fsExtra().default.readFile(_path().default.join(buildGradlePath), 'utf-8');
    const applyEasGradle = 'apply from: "./eas-build.gradle"';
    const hasEasGradleApply = buildGradleContent.split('\n') // Check for both single and double quotes
    .some(line => line === applyEasGradle || line === applyEasGradle.replace(/"/g, "'"));
    return hasEasGradleApply && hasEasGradleFile;
  }

  async ensureProjectConfiguredAsync() {
    if (!(await this.isProjectConfiguredAsync())) {
      throw new (_CommandError().default)('Project is not configured. Please run "expo eas:build:init" first to configure the project');
    }
  }

  async configureProjectAsync() {
    const spinner = (0, _ora().default)('Making sure your Android project is set up properly');

    if (await this.isProjectConfiguredAsync()) {
      spinner.succeed('Android project is already configured');
      return;
    }

    const {
      projectDir
    } = this.ctx.commandCtx;

    const androidAppDir = _path().default.join(projectDir, 'android', 'app');

    const buildGradlePath = _path().default.join(androidAppDir, 'build.gradle');

    const easGradlePath = _path().default.join(androidAppDir, 'eas-build.gradle');

    await _fsExtra().default.writeFile(easGradlePath, _gradleContent().default);
    await (0, _git().gitAddAsync)(easGradlePath, {
      intentToAdd: true
    });
    const buildGradleContent = await _fsExtra().default.readFile(_path().default.join(buildGradlePath), 'utf-8');
    const applyEasGradle = 'apply from: "./eas-build.gradle"';
    await _fsExtra().default.writeFile(buildGradlePath, `${buildGradleContent.trim()}\n${applyEasGradle}\n`);

    try {
      await gitUtils().ensureGitStatusIsCleanAsync();
      spinner.succeed();
    } catch (err) {
      if (err instanceof gitUtils().DirtyGitTreeError) {
        spinner.succeed('We configured your Android project to build it on the Expo servers');

        _log().default.newLine();

        try {
          await gitUtils().reviewAndCommitChangesAsync('Configure Android project', {
            nonInteractive: this.ctx.commandCtx.nonInteractive
          });
          (0, _log().default)(`${_chalk().default.green(_figures().default.tick)} Successfully committed the configuration changes.`);
        } catch (e) {
          throw new Error("Aborting, run the command again once you're ready. Make sure to commit any changes you've made.");
        }
      } else {
        spinner.fail();
        throw err;
      }
    }
  }

  async prepareJobAsync(archiveUrl) {
    if (!this.credentialsPrepared) {
      throw new Error('ensureCredentialsAsync should be called before prepareJobAsync');
    }

    if (this.ctx.buildProfile.workflow === _easJson().Workflow.Generic) {
      return (0, _easBuildJob().sanitizeJob)((await this.prepareGenericJobAsync(archiveUrl, this.ctx.buildProfile)));
    } else if (this.ctx.buildProfile.workflow === _easJson().Workflow.Managed) {
      return (0, _easBuildJob().sanitizeJob)((await this.prepareManagedJobAsync(archiveUrl, this.ctx.buildProfile)));
    } else {
      throw new Error("Unknown workflow. Shouldn't happen");
    }
  }

  async prepareJobCommonAsync(archiveUrl) {
    const buildCredentials = this.credentials ? {
      buildCredentials: {
        keystore: {
          dataBase64: this.credentials.keystore.keystore,
          keystorePassword: this.credentials.keystore.keystorePassword,
          keyAlias: this.credentials.keystore.keyAlias,
          keyPassword: this.credentials.keystore.keyPassword
        }
      }
    } : {};
    return {
      platform: _easBuildJob().Platform.Android,
      projectUrl: archiveUrl,
      secrets: { ...(this.secretEnvs ? {
          secretEnvs: this.secretEnvs
        } : {}),
        ...buildCredentials
      }
    };
  }

  async prepareGenericJobAsync(archiveUrl, buildProfile) {
    const projectRootDirectory = _path().default.relative((await (0, _git().gitRootDirectory)()), process.cwd()) || '.';
    return { ...(await this.prepareJobCommonAsync(archiveUrl)),
      type: _easBuildJob().BuildType.Generic,
      gradleCommand: buildProfile.gradleCommand,
      artifactPath: buildProfile.artifactPath,
      projectRootDirectory
    };
  }

  async prepareManagedJobAsync(archiveUrl, _buildProfile) {
    return { ...(await this.prepareJobCommonAsync(archiveUrl)),
      type: _easBuildJob().BuildType.Managed,
      packageJson: {
        example: 'packageJson'
      },
      manifest: {
        example: 'manifest'
      }
    };
  }

  shouldLoadCredentials() {
    return this.ctx.buildProfile.workflow === _easJson().Workflow.Managed || this.ctx.buildProfile.workflow === _easJson().Workflow.Generic && !this.ctx.buildProfile.withoutCredentials;
  }

}

var _default = AndroidBuilder;
exports.default = _default;
